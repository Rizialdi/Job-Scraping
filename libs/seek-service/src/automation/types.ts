/**
 * Parse the data-search-sol-meta attribute from the job ad
 */
interface IDataSearchSolMeta {
  searchRequestToken: '333bc538-79f2-4e09-b6f2-60afa8ebee57';
  token: '0~333bc538-79f2-4e09-b6f2-60afa8ebee57';
  jobId: '59820105';
  section: 'MAIN';
  sectionRank: 22;
  jobAdType: 'ORGANIC';
  tags: {
    mordor__flights: 'mordor_123';
    mordor__s: '0';
    'seek:chalice:experiments:job_details_split_view': '2';
    'seek:chalice:experiments:aa_test': '2';
  };
}

/**
 * Job listing interface from scraping the Seek website
 */
interface IJobListing {
  companyName: string;
  jobUrl: string;
  jobId: string;
  location: string;
  jobTitle: string;
  jobListingDate: string;
  jobClassification: string;
  jobSpecifics: string;
  jobSalaryRange: string;
  jobContractType: string;
  jobSalaryType: string;
  jobAdType: string;
  roleId: string;
  logo: string;
  jobTeaser: string;
}

export { IDataSearchSolMeta, IJobListing };
