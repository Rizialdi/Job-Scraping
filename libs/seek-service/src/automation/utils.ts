import puppeteer from 'puppeteer';
import { IDataSearchSolMeta, IJobListing } from './types';

/**
 * Search for a keyword on Seek website
 * @param keyword - The keyword to search for
 * @returns The list of job listings
 */
const search = async (keyword: string): Promise<IJobListing[]> => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  const url = 'https://www.seek.com.au';

  // Navigate to the website
  await page.goto(url);

  // Set screen size
  await page.setViewport({ width: 1080, height: 1024 });

  // Find the search box
  const keywordAnchor = await page.$("input[placeholder='Enter Keywords']");
  // Find the search button
  const searchAnchor = await page.$("button[id='searchButton']");

  if (!keywordAnchor) throw new Error('No anchor found');
  if (!searchAnchor) throw new Error('No anchor found');

  // Type into search box
  await keywordAnchor.type(keyword);
  // Click on search button
  await searchAnchor.click();

  // Wait for the results to load
  await page.waitForSelector('div._1wkzzau0.a1msqi6j > div._1wkzzau0');

  // Iterate over the results and extract the data
  const jobListings = await page.evaluate((url) => {
    return Array.from(
      document.querySelectorAll('div._1wkzzau0.a1msqi6j > div._1wkzzau0')
    ).map((jobListing) => {
      const meta = jobListing
        .querySelector('div > div')
        ?.getAttribute('data-search-sol-meta');

      const { jobId, jobAdType }: IDataSearchSolMeta = JSON.parse(meta);

      return {
        companyName:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > div > span > a'
          )?.textContent ?? 'Private Advertiser',
        jobUrl:
          url + jobListing.querySelector('div > a')?.getAttribute('href') ?? '',
        jobId: jobId,
        location:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > span > a'
          )?.textContent ?? '',
        jobTitle:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > div > div > h3 > a'
          )?.textContent ?? '',
        jobListingDate:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > span > span'
          )?.textContent ?? '',
        jobClassification:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > span > div > div > div > a'
          )?.textContent ?? '',
        jobSpecifics:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > span'
          )?.textContent ?? '',
        jobSalaryRange:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > div > div > div > span > div'
          )?.textContent ?? '',
        jobContractType:
          jobListing.querySelector('div > div> div > div > div > div > p')
            ?.textContent ?? '',
        jobSalaryType: '',
        jobAdType: jobAdType,
        roleId: '',
        logo:
          jobListing
            .querySelector('div > div > div > div > div > img')
            ?.getAttribute('src') ?? '',
        jobTeaser:
          jobListing.querySelector(
            'div > div > article > div > div > div > div > div > span'
          )?.textContent ?? '',
      };
    });
  }, url);

  await browser.close();

  return jobListings;
};

export { search };
