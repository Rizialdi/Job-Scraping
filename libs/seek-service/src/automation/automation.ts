import { IJobListing } from './types';
import { search } from './utils';

/**
 * SeekService Automation
 * This class is responsible for scraping the Seek website
 */
class SeekServiceAutomation {
  /**
   * Get job listings given a keyword
   * @param keyword - The keyword to search for
   * @returns The response from the API
   */
  static async search(keyword: string): Promise<IJobListing[]> {
    return search(keyword);
  }
}

export { SeekServiceAutomation };
