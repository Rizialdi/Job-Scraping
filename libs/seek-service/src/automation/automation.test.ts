import { searchResult } from '../__tests__/stubs/automation.stub';
import { SeekServiceAutomation } from './automation';

describe('#SeekServiceAutomation', () => {
  let keyword: string;
  let sut: SeekServiceAutomation;

  beforeEach(() => {
    keyword = 'fullstack';
    sut = new SeekServiceAutomation();
  });

  describe('#constructor', () => {
    it('should be defined', () => {
      expect(sut).toBeDefined();
    });
  });

  describe('#search', () => {
    it('should call the correct method', async () => {
      const mockedSearch = jest.spyOn(sut, 'search').mockResolvedValue([]);

      await sut.search(keyword);

      expect(mockedSearch).toHaveBeenCalledTimes(1);
      expect(mockedSearch).toHaveBeenCalledWith(keyword);
    });

    it('should have the response matching the snapshot', async () => {
      jest.spyOn(sut, 'search').mockResolvedValue(searchResult());

      const response = await sut.search(keyword);

      expect(response).toMatchSnapshot();
    });
  });
});
