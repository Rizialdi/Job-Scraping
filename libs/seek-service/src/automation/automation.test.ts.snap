// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#SeekServiceAutomation #search should have the response matching the snapshot 1`] = `
[
  {
    "companyName": "Evolution Recruitment Solutions Pty Ltd",
    "jobAdType": "ORGANIC",
    "jobClassification": "Information & Communication Technology",
    "jobContractType": "This is a Full Time job",
    "jobId": "59820105",
    "jobListingDate": "4 Feb 2023",
    "jobSalaryRange": "$140,000 - $169,999",
    "jobSalaryType": "",
    "jobSpecifics": "Your team will be responsible building the core product and be the Engine behind the success of Unicorn Tech Disruptor.",
    "jobTeaser": "Your team will be responsible building the core product and be the Engine behind the success of Unicorn Tech Disruptor.",
    "jobTitle": "Senior  Fullstack .NET Software Engineer",
    "jobUrl": "https://www.seek.com.au/job/59820105?type=standout#sol=09b8a51e2d80eebd75c222500c3ad8ff7159f322",
    "location": "Sydney NSW",
    "logo": "",
    "roleId": "",
  },
]
`;
